File: 2181-merge-nodes-in-between-zeros/2181-merge-nodes-in-between-zeros.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

struct ListNode* getNode(int x)
{
    struct ListNode* newNode = malloc(sizeof(struct ListNode));
    newNode->val = x;
    newNode->next = NULL;
    return newNode;
}

struct ListNode* mergeNodes(struct ListNode* head){
    struct ListNode* h1;
    struct ListNode* tmp;
    int sum = 0;
    int c = 0;
    
    
    while(head != NULL)
    {
        if(head->val == 0)
        {
            if(sum != 0)
            {
                if(c == 0)   //only once
                { 
                    h1 = getNode(sum);
                    tmp = h1;
                    c = 1;
                }
                else
                {
                    tmp->next = getNode(sum);
                    tmp = tmp->next;
                }
            }
            sum = 0;
        }
        else
        {
            sum += head->val;
        }
        head = head->next;
    }
    
    return h1;
}

File: 2181-merge-nodes-in-between-zeros/2181-merge-nodes-in-between-zeros.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeNodes(ListNode head) {
        ListNode list = null;
        ListNode tmp = null;
        
        int sum = 0;
        head = head.next;
        while(head != null)
        {
            if(head.val == 0)
            {
                ListNode node = new ListNode(sum);
                if(list == null)
                {
                    list = node;
                    tmp = list;
                }
                else{
                    tmp.next = node;
                    tmp = tmp.next;
                }
                sum = 0;
            }
            else
                sum += head.val;
            head = head.next;
        }
        return list;
    }
}